### DISCUSSION TITLE
Using XState for implementing a server-side order workflow

### DISCUSSION THREAD

**Discussion by KevinGaudin** [statelyai/xstate#1684] [Q&A]
Hi,

I'm looking for some TypeScript libraries that would help implementing a server-side order management workflow in a nodejs server application.

XState seems to be popular in the Node community but most use cases seem to be about UX implementation.

Is XState appropriate for handling persisted objects in a DB and make them follow the state-machine workflow based on external events and trigger external actions ?

Thanks for your insights !
Kevin

**Comment by davidkpiano** [comment]
> Is XState appropriate for handling persisted objects in a DB and make them follow the state-machine workflow based on external events and trigger external actions?

Yes! Check out this example using [Azure Durable Entities](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-entities) with XState in a stateful, serverless environment: https://github.com/davidkpiano/durable-entities-xstate

**Reply by KevinGaudin** [reply]
Thanks for your fast answer, sounds cool, I'll have a look.

**Reply by markwallsgrove** [reply]
@davidkpiano Thank you for your example. As for general advise, should the state be checked within the database before persisting a new state? i.e. the persisted state might have been mutated while the current process is running.

**Reply by davidkpiano** [reply]
> @davidkpiano Thank you for your example. As for general advise, should the state be checked within the database before persisting a new state? i.e. the persisted state might have been mutated while the current process is running.

Well, with a queue (how DF works internally), you won't have that problem since anything else that can possibly modify state is stuck in the queue while the current "request" is being processed, so you don't have race conditions.

**Reply by markwallsgrove** [reply]
Ah, I see. So there has to be a single instance processing a queue of requests? How would this work if say the state was related to a mortgage (where the lifetime is in years) and there are many instances (maybe for a mortgage company)?

I assumed that I could recreate the state machine based on the state within the database, apple the action, and then persist the new state / values if the state within the database was the same value as the original value. 

Maintaining x processes seems impractical where x is more than a trival amount.

**Reply by davidkpiano** [reply]
Well, let's say you have a mortgage workflow. An _instance_ of that mortgage workflow should handle one request at a time, so you would have a queue for that workflow instance.

If you have _many instances_, you would have a queue for each of those instances (so, multiple queues). No scalability issues there.

**Reply by markwallsgrove** [reply]
Oh, OK. This sounds a lot like the issue that AWS Kinesis solves.

How would you solve this issue without any proprietary technology such as Azure Durable Entities or AWS Kinesis? Providing a standalone example might make the solution easier to understand.

Having a queue per mortgage agreement seems incredibly expensive from a management and cost point of view.  Adding a queue into the mix where the results need to be sent to the UI also seems to add complications.

My idea to solve these issues (from a naive point of view) was to:
- within the api request load the state machine from the database
- apply the required action and manipulate the data
- Execute a database query which would update the row by id where the state was the original value
- if the state had changed reload the state machine from the new state & try again (retry multiple times if required)
- respond with a payload that reflects the end result (successful / failed)

This would allow the api deployment to be stateless & the database would act as a way of ensuring the state hadn't changed while the state machine was processing the request.

For an action that requires a async change I did consider if a new state could be used.
- within the api request load the state machine from the database
- apply a action that changes the state to ready the machine for a async change (persist change to db)
- if the state change was successful run the async change & modify the status back to the original value (if desired)
- retry with new state if the second stat wasn't successful for x times

There are a few benefits to this approach such as:
- easy to respond to a request with the end result (this would be hard/complicated to solve using a queue approach)
- database acts as a locking mechanism to restrict the changes across multiple requests
- requires no extra ram
- doesn't need to be maintained

Confusingly this approach doesn't seem to match your expectations of how to use the library on the server side.

**Reply by hmvs** [reply]
Hey @markwallsgrove have you found anything that matches your use case? I have a similar use case and looking for library that could help me with that.

**Reply by markwallsgrove** [reply]
👋  @hmvs. Sadly I didn't find a conclusion to this investigation. My best guess would be to use AWS SQS FIFO queues where the message group id is set to something like `${objectType}:${objectId}`. The queue could then be processed with a AWS Lambda function, which is where XState would be used.

**Reply by hmvs** [reply]
I see. Thanks, I thought you might have found another library.

**Reply by markwallsgrove** [reply]
The crooks of my argument `
Having a queue per mortgage agreement seems incredibly expensive from a management and cost point of view.` isn't true if the AWS SQS FIFO is used with a message group id.

Waiting for a response is complicated though (when it's required). Currently others have used the approach of creating a new SQS queue with a predictable name & adding a new item to that queue once the request has been processed. Amazon has created a temporary queue mechanism, but it's only available in Java :(.

**Reply by gaffarmalik** [reply]
A good idea here is to stick to your database. 
Have some worker pick jobs from the db.
Lock a row that's currently in use by some other wrkflow instance and enjoy your day.
